src/mlpack/bindings/java/CMakeLists.txt

```diff
@@ -0,0 +1,227 @@
+    WORKING_DIRECTORY ${MAVEN_BUILD_DIR})
+
+# Javacpp was pulled from maven
+set(JAVACPP ${MAVEN_BUILD_DIR}/lib/javacpp-${JAVACPP_VERSION}.jar CACHE INTERNAL "")
+
+# Compile cli_util library for java
+add_custom_target(jniCLI_configure ALL
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_SOURCE_DIR}/maven/CLI.java
+        ${MAVEN_BUILD_DIR}/src/main/java/org/mlpack/CLI.java
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_SOURCE_DIR}/maven/MatrixWithInfo.java
+        ${MAVEN_BUILD_DIR}/src/main/java/org/mlpack/MatrixWithInfo.java
+    DEPENDS java_configure)
+
+add_custom_command(
+    OUTPUT
+        ${BUILD_SRC_DIR}/cli_util.hpp
+        ${BUILD_SRC_DIR}/deleter.hpp
+        ${BUILD_SRC_DIR}/cli_util.cpp
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_SOURCE_DIR}/cli_util.hpp
+        ${BUILD_SRC_DIR}/cli_util.hpp
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_SOURCE_DIR}/deleter.hpp
+        ${BUILD_SRC_DIR}/deleter.hpp
+    COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_SOURCE_DIR}/cli_util.cpp
+        ${BUILD_SRC_DIR}/cli_util.cpp)
+
+# Compile java sources with maven
+add_custom_command(TARGET jniCLI_configure PRE_BUILD
+    COMMAND ${MAVEN} compile -Dchanged.classes=org/mlpack/CLI.java
+    WORKING_DIRECTORY ${MAVEN_BUILD_DIR})
+
+# Generate JNI wrappers from compiled java bytecode
+add_custom_command(OUTPUT ${JNI_GEN_DIR}/jnijavacpp.cpp ${JNI_GEN_DIR}/jniCLI.cpp
+    COMMAND ${Java_JAVA_EXECUTABLE} -jar ${JAVACPP} -classpath .:${MAVEN_BUILD_DIR}/lib/* -nocompile -d ${JNI_GEN_DIR} org/mlpack/CLI.class
+    WORKING_DIRECTORY ${MAVEN_BUILD_DIR}/target/classes
+    DEPENDS jniCLI_configure)
+
+add_library(jnijavacpp STATIC
+    ${JNI_GEN_DIR}/jnijavacpp.cpp)
+
+target_include_directories(jnijavacpp PUBLIC ${JNI_INCLUDE})
+set_target_properties(jnijavacpp PROPERTIES
+    COMPILE_FLAGS -fPIC)
+
+# Compile shared object
+add_library(jniCLI SHARED
+    ${JNI_GEN_DIR}/jniCLI.cpp
+    ${BUILD_SRC_DIR}/cli_util.hpp
+    ${BUILD_SRC_DIR}/cli_util.cpp
+    ${BUILD_SRC_DIR}/deleter.hpp)
+
+target_include_directories(jniCLI PUBLIC ${BUILD_SRC_DIR} ${JNI_GEN_DIR} ${JNI_INCLUDE})
+target_link_libraries(jniCLI mlpack jnijavacpp)
+set_target_properties(jniCLI PROPERTIES
+    LIBRARY_OUTPUT_DIRECTORY "${NATIVE_OUT_DIR}")
+add_dependencies(jniCLI jniCLI_configure)
+add_dependencies(java jniCLI)
+
+add_library(java_generate_parts STATIC
+    ${CMAKE_CURRENT_SOURCE_DIR}/strip_type.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_param_defn.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_output_param.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_java.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_java.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_input_param.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/print_doc_functions.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/java_option.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/get_java_type.hpp)
+
+macro(add_java_binding file_name class_name)
```

Comment:
Agreed, the reason for all these arguments is a Java rule that file name and class name have to be the same. Thus, even though we are compiling `knn` method in the example above, its corresponding Java class will be called `Knn` according to Java style rules. I simply didn't want to get my hands dirty with regexp here ;). Still, it will be fixed when I push new binding generation process I'm working on right now

Reply:
Fixed in 81ab8ee